File: kotlin/wordy/src/main/kotlin/Wordy.kt
import kotlin.math.pow

object Wordy {

    fun answer(input: String): Int =
        extractTokens(input).let { tokens ->
            // we should start with a number, and then alternate operators and numbers
            require(tokens.size % 2 == 1)
            var result = tokens[0].toIntOrNull(10) ?: throw IllegalArgumentException()
            for (i in 1 until tokens.size step 2) {
                val op = tokens[i]
                val num = tokens[i + 1].toIntOrNull(10) ?: throw IllegalArgumentException()
                when (op) {
                    "plus" -> result += num
                    "minus" -> result -= num
                    "multiplied" -> result *= num
                    "divided" -> result /= num
                    "pow" -> result = result.toDouble().pow(num).toInt()
                    else -> throw IllegalArgumentException()
                }
            }
            result
        }

    private fun extractTokens(input: String): List<String> =
        (Regex("What is (.+)\\?").matchEntire(input) ?: throw IllegalArgumentException())
            .groupValues[1]
            .replace(Regex("(multiplied|divided) by")) { it.groupValues[1] }
            .replace(Regex("raised to the (\\d+)(?:st|nd|rd|th) power")) { "pow ${it.groupValues[1]}" }
            .split(' ')
}

File: elixir/wordy/lib/wordy.ex
defmodule Wordy do
  @type op :: (integer, integer -> integer) | nil

  @doc """
  Calculate the math problem in the sentence.
  """
  @spec answer(String.t()) :: integer
  def answer(question) do
    words =
      question
      |> String.downcase()
      |> String.replace(~r/^\s*what is/, "")
      |> String.replace(~r/[?]\s*$/, "")
      |> String.split()

    # start with a number
    get_number(words)
  end

  # ------------------------------------------------------------
  # The next word _should_ be a number.
  # - &String.to_integer/1 raises ArgumentError

  defp get_number(words, operation \\ nil, accumulator \\ 0)

  # Can't end an expression seeking the next number
  defp get_number([], _, _),
    do: raise(ArgumentError)

  # Initial iteration passes nil as the operation
  # Set the accumulator to the value of the first number
  defp get_number([num | words], nil, _),
    do: get_operation(words, String.to_integer(num))

  defp get_number([num | words], operation, acc),
    do: get_operation(words, operation.(acc, String.to_integer(num)))

  # ------------------------------------------------------------
  # The next word(s) _should_ be an arithmetic operation.

  # When there are no more words, we have the answer
  defp get_operation([], answer),
    do: answer

  defp get_operation(["plus" | words], acc),
    do: get_number(words, &Kernel.+/2, acc)

  defp get_operation(["minus" | words], acc),
    do: get_number(words, &Kernel.-/2, acc)

  defp get_operation(["multiplied", "by" | words], acc),
    do: get_number(words, &Kernel.*/2, acc)

  defp get_operation(["divided", "by" | words], acc),
    do: get_number(words, &Kernel.div/2, acc)

  defp get_operation(_, _),
    do: raise(ArgumentError)
end
